### PR TITLE
Replaced hardcoded custom option types and groups with injected array…

### PR DESCRIPTION
**Pull request by zhartaunik**
… arguments (#24726)

### Description (*)

Changes are helping to improve the scaleability of the Magento custom options list.

### Fixed Issues (if relevant)

1. magento/magento2#24726: Custom option type factory is bound to a folder

### Manual testing scenarios (*)

1. Create a product with custom options (better add with all types, as we modified all custom options creation)
2. Try to add this product to a shopping cart
ER: Product should be added with all custom options without any errors

### Contribution checklist (*)
 - [*] Pull request has a meaningful description of its purpose
 - [*] All commits are accompanied by meaningful commit messages
 - [*] All new or changed code is covered with unit/integration tests (if applicable)
 - [*] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by zhartaunik** [comment]
Friends also please advice, if any code refactoring in changed files will be appreciated. (In my case ```Catalog/etc/di.xml```, ```\Magento\Catalog\Model\Product\Option```)
For example import classes, removing obsolete injections from di.xml, deeper refactoring like expelling methods, etc.

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 1]
```diff
@@ -306,44 +323,30 @@ public function setProduct(Product $product = null)
...
-            self::OPTION_TYPE_RADIO => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_CHECKBOX => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_MULTIPLE => self::OPTION_GROUP_SELECT,
-            self::OPTION_TYPE_DATE => self::OPTION_GROUP_DATE,
-            self::OPTION_TYPE_DATE_TIME => self::OPTION_GROUP_DATE,
-            self::OPTION_TYPE_TIME => self::OPTION_GROUP_DATE,
-        ];
 
-        return $optionGroupsToTypes[$type] ?? '';
+        return $this->typesPool[$type] ?? '';
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: Using "pool" in the property name will be confusing there because of https://en.wikipedia.org/wiki/Object_pool_pattern.

Initial variable name $optionGroupsToTypes was much clear. Could you use it for property name?
Could you adjust also property name for groups?

Related issue: https://github.com/magento/architecture/issues/154

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 2]
```diff
@@ -306,44 +323,30 @@ public function setProduct(Product $product = null)
...
     /**
      * Group model factory
      *
      * @param string $type Option type
-     * @return \Magento\Catalog\Model\Product\Option\Type\DefaultType
+     * @return DefaultType
      * @throws LocalizedException
      */
-    public function groupFactory($type)
+    public function groupFactory($type): DefaultType
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: As you can see - earlier it was using all classes from namespace `Magento\Catalog\Model\Product\Option\Type\\` - so potentially not all could be extended from `\Magento\Catalog\Model\Product\Option\Type\DefaultType` - as result it might cause type error here.

Could you double check that?

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 3]
```diff
@@ -135,12 +148,16 @@ public function __construct(
         \Magento\Framework\Model\ResourceModel\AbstractResource $resource = null,
         \Magento\Framework\Data\Collection\AbstractDb $resourceCollection = null,
         array $data = [],
+        array $groupsPool = [],
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: All new constructor arguments should be added at the end. Please fix that

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@zhartaunik Also please check unit test failures and adjust your code/unit test accordingly

File: app/code/Magento/Catalog/Model/Product/Option.php (hunk 1 above)
Comment (zhartaunik) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Agree

File: app/code/Magento/Catalog/Model/Product/Option.php (hunk 2 above)
Comment (zhartaunik) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Original class returned DefaultType. It already have too many methods which are required for initialization and rendering custom options in different places. And this DefaultType does not have an interface. It looks like some abstraction, but there no word "Abstract".
I agree with you that we tried to make better for users and at the same time put such restriction. Maybe yes, it will be better to remove strict return, but what to put inside PHPDoc @return? DataObject?

File: app/code/Magento/Catalog/Model/Product/Option.php (hunk 3 above)
Comment (zhartaunik) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Affirmative

**Comment by zhartaunik** [comment]
> @zhartaunik Also please check unit test failures and adjust your code/unit test accordingly

Yep, I'll check. @ihor-sviziev Please help with the question posted early:

> Friends also please advice, if any code refactoring in changed files will be appreciated. (In my case Catalog/etc/di.xml, \Magento\Catalog\Model\Product\Option)
For example import classes, removing obsolete injections from di.xml, deeper refactoring like expelling methods, etc.

**Comment by zhartaunik** [comment]
Can anybody advice, what does mean "Semantic Version Checker error"? Modules do not have versions at all. Neither in composer.json and in module.xml

File: app/code/Magento/Catalog/etc/di.xml [hunk 4]
```diff
@@ -72,7 +72,6 @@
     <preference for="Magento\Catalog\Model\Indexer\Product\Price\UpdateIndexInterface" type="Magento\Catalog\Model\Indexer\Product\Price\InvalidateIndex" />
     <preference for="Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface" type="Magento\Catalog\Model\Product\Gallery\ImagesConfigFactory" />
     <preference for="Magento\Catalog\Model\Product\Configuration\Item\ItemResolverInterface" type="Magento\Catalog\Model\Product\Configuration\Item\ItemResolverComposite" />
-    <preference for="Magento\Catalog\Api\Data\MassActionInterface" type="\Magento\Catalog\Model\MassAction" />
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/etc/di.xml]: Please revert this change, it’s not related to your changes

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 5]
```diff
@@ -306,44 +323,30 @@ public function setProduct(Product $product = null)
...
-        return $optionGroupsToTypes[$type] ?? '';
+        return $this->optionGroupsToTypes[$type] ?? '';
     }
 
     /**
      * Group model factory
      *
      * @param string $type Option type
-     * @return \Magento\Catalog\Model\Product\Option\Type\DefaultType
+     * @return DataObject
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: Please revert this change, let’s leave `\Magento\Catalog\Model\Product\Option\Type\DefaultType` there

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @zhartaunik,
Semantic module version checks that public method arguments and return types are not changes as it would be treated as backward incompatible change. I just wrote comments what should be reverted, actually in one place you fixed argument type, and there it’s ok

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 6]
```diff
@@ -135,14 +148,18 @@ public function __construct(
         \Magento\Framework\Model\ResourceModel\AbstractResource $resource = null,
         \Magento\Framework\Data\Collection\AbstractDb $resourceCollection = null,
         array $data = [],
-        ProductCustomOptionValuesInterfaceFactory $customOptionValuesFactory = null
+        ProductCustomOptionValuesInterfaceFactory $customOptionValuesFactory = null,
+        array $optionGroups = [],
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php:156]: Actually someone could override this class in own module and not pass these two arguments, in this case - we'll have empty array here. please add default value, similar to getting objects from object manager, but in this case it will be default array

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/Product/Option.php (hunk 6 above)
Comment (zhartaunik) [reply app/code/Magento/Catalog/Model/Product/Option.php:156]: Understood. Sorry for such issues, I had no experience in core contribution. With it, I reverted unit tests, as a model now has backward compatibility.

File: app/code/Magento/Catalog/Model/Product/Option.php (hunk 5 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Still not fixed, please check it

Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Sorry, just looked at old code

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 7]
```diff
@@ -135,14 +148,34 @@ public function __construct(
...
     ) {
         $this->productOptionValue = $productOptionValue;
         $this->optionTypeFactory = $optionFactory;
-        $this->validatorPool = $validatorPool;
         $this->string = $string;
+        $this->validatorPool = $validatorPool;
         $this->customOptionValuesFactory = $customOptionValuesFactory ?:
             \Magento\Framework\App\ObjectManager::getInstance()->get(ProductCustomOptionValuesInterfaceFactory::class);
+        $this->optionGroups = $optionGroups ?: [
+            'date' => 'Magento\Catalog\Model\Product\Option\Type\Date',
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: 1. Please use `::class` for class names
2. I guess it’s possible to use constants there. Could you use them, or am I wrong?

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 8]
```diff
@@ -135,14 +148,34 @@ public function __construct(
...
+            'file' => 'Magento\Catalog\Model\Product\Option\Type\File',
+            'select' => 'Magento\Catalog\Model\Product\Option\Type\Select',
+            'text' => 'Magento\Catalog\Model\Product\Option\Type\Text',
+        ];
+        $this->optionGroupsToTypes = $optionGroupsToTypes ?: [
+            'field' => 'text',
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: Please use constants there, as it was before

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi,
Just few really small issues found, everything else looks good

File: app/code/Magento/Catalog/Model/Product/Option.php (hunk 7 above)
Comment (zhartaunik) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Added constants here and inside di.xml too

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 9]
```diff
@@ -135,14 +152,34 @@ public function __construct(
...
+        $this->validatorPool = $validatorPool;
         $this->customOptionValuesFactory = $customOptionValuesFactory ?:
             \Magento\Framework\App\ObjectManager::getInstance()->get(ProductCustomOptionValuesInterfaceFactory::class);
+        $this->optionGroups = $optionGroups ?: [
+            self::OPTION_GROUP_DATE => Date::class,
+            self::OPTION_GROUP_FILE => File::class,
+            self::OPTION_GROUP_SELECT => Select::class,
+            self::OPTION_GROUP_TEXT => Text::class,
+        ];
+        $this->optionGroupsToTypes = $optionGroupsToTypes ?: [
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Option.php]: Once you added constants - it’s became clear that it’s actually opposite mapping, option type to group, please rename the property

Sorry that didn’t realized it earlier

Comment (zhartaunik) [reply app/code/Magento/Catalog/Model/Product/Option.php]: Also sorry for my inattention. You spent on my PR too much time.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6119](https://jira.corp.magento.com/browse/ENGCOM-6119) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@zhartaunik thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed
